### PR TITLE
[heft] Fix issues related to running multiple Heft sessions in a single process

### PR DESCRIPTION
**Pull request by dmichon-msft**
## Summary
Fixes various bugs encountered while trying to reuse a single Heft process for multiple sessions (to reduce boot overhead in a large monorepo).

## Details
Fixed mutation of loaded config file objects that led to JSON Schema violations if the same file was requested a second time, or in an invalid config state.

Moved validation of plugin name collisions to the InternalHeftSession initialization, so that collisions are only detected within the scope of a single session.

Reduced verbosity of the "this plugin is not supported in watch mode" messages for Lint and API-Extractor to only log on the initial run.

Added an option to API-Extractor to make it run even in watch mode (by performing a full run).

Updated the Jest plugin to wait for the initial test run on the first iteration.

## How it was tested
Validated that `rush retest` in rushstack repository still works as expected.
Linked against prototype of Heft process reuse.

## Impacted documentation
None.

### CODE REVIEW

File: apps/heft/src/configuration/HeftPluginDefinition.ts [hunk 1]
```diff
@@ -216,17 +214,6 @@ export abstract class HeftPluginDefinitionBase {
 
     // Ensure that plugin names are unique. Main reason for this restriction is to ensure that command-line
     // parameter conflicts can be handled/undocumented synonyms can be provided in all scenarios
```
Comment (iclanton) [inline apps/heft/src/configuration/HeftPluginDefinition.ts:216]: Remove this comment?

File: heft-plugins/heft-api-extractor-plugin/src/ApiExtractorPlugin.ts [hunk 2]
```diff
@@ -165,19 +172,25 @@ export default class ApiExtractorPlugin implements IHeftTaskPlugin {
...
-
     const apiExtractorTaskConfiguration: IApiExtractorTaskConfiguration | undefined =
       await this._getApiExtractorTaskConfigurationAsync(taskSession, heftConfiguration);
 
+    if (runOptions.requestRun) {
+      if (!apiExtractorTaskConfiguration?.runInWatchMode) {
+        if (!this._printedWatchWarning) {
+          this._printedWatchWarning = true;
+          taskSession.logger.terminal.writeWarningLine(
+            "API Extractor isn't currently supported in -watch mode."
```
Comment (iclanton) [inline heft-plugins/heft-api-extractor-plugin/src/ApiExtractorPlugin.ts:187]:
```suggestion
            "API Extractor isn't currently enabled in watch mode."
```

File: heft-plugins/heft-typescript-plugin/src/TypeScriptBuilder.ts [hunk 3]
```diff
@@ -350,12 +362,21 @@ export class TypeScriptBuilder {
...
       measureSync: measureTsPerformance,
       pendingOperations,
       rawDiagnostics,
       pendingTranspilePromises
     } = tool;
 
+    // Remove this when watchFile and watchDirectory have been overridden
+    if (ts.sys.getCurrentDirectory() !== system.getCurrentDirectory()) {
+      // This happens because ts.sys.watchFile and ts.sys.watchDirectory wrap a memoized working directory.
+      throw new Error(
```
Comment (iclanton) [inline heft-plugins/heft-typescript-plugin/src/TypeScriptBuilder.ts:375]: So will this just cause Heft to crash in some cases?

**Review by iclanton — APPROVED** [review]

File: heft-plugins/heft-typescript-plugin/src/TypeScriptBuilder.ts (hunk 3 above)
Comment (dmichon-msft) [reply heft-plugins/heft-typescript-plugin/src/TypeScriptBuilder.ts:375]: On further inspection, this mismatch shouldn't actually break anything, so I can remove this check.

File: heft-plugins/heft-typescript-plugin/src/TypeScriptBuilder.ts [hunk 4]
```diff
@@ -350,12 +362,21 @@ export class TypeScriptBuilder {
...
+        `TypeScript has internally memoized an incorrect working directory in the watcher. Watch mode cannot function properly.`
+      );
+    }
+
```
Comment (dmichon-msft) [inline heft-plugins/heft-typescript-plugin/src/TypeScriptBuilder.ts:379]:
```suggestion
```

File: apps/heft/src/configuration/HeftPluginDefinition.ts (hunk 1 above)
Comment (dmichon-msft) [inline apps/heft/src/configuration/HeftPluginDefinition.ts:216]:
```suggestion
```

File: heft-plugins/heft-typescript-plugin/src/TypeScriptBuilder.ts [hunk 5]
```diff
@@ -350,6 +362,7 @@ export class TypeScriptBuilder {
   public async _runWatchAsync(tool: ITypeScriptTool): Promise<void> {
     const {
       ts,
+      system,
```
Comment (dmichon-msft) [inline heft-plugins/heft-typescript-plugin/src/TypeScriptBuilder.ts:365]:
```suggestion
```

**Review by iclanton — APPROVED** [review]

**Pull request merged**